### PR TITLE
Query for topics metrics

### PR DESCRIPTION
**Pull request by faderskd**
This pull request adds additional endpoint (/query/topics/metrics) which handle POST method for fetching list of topics metrics. Two new options for filtering data were added. They are GRATER_THAN and LOWER_THAN operator. Simple example for filtering topics metrics using one of them:

POST /query/topics/metrics
{
    "query": {"rate": {"gt": 0}}
}

### CODE REVIEW

File: integration/src/integration/java/pl/allegro/tech/hermes/integration/management/QueryEndpointTest.java [hunk 1]
```diff
@@ -143,6 +143,37 @@ public void shouldSkipEntitiesNotContainingQueriedNestedField() {
...
+                {"{\"query\": {\"published\": {\"gt\": \"5\"}}}", asList()},
+                {"{\"query\": {\"published\": {\"gt\": \"2\"}}}", asList("testGroup.testTopic2")},
+                {"{\"query\": {\"published\": {\"lt\": \"2\"}}}", asList("testGroup.testTopic1")},
+        };
+    }
+
+    @Test(dataProvider = "topicsMetricsFilteringData")
+    public void shouldQueryTopicsMetrics(String query, List<String> qualifiedNames) {
+        // given
+        operations.buildTopic(topic("testGroup", "testTopic1").withContentType(JSON).withTrackingEnabled(false).build());
```
Comment (adamdubiel) [inline integration/src/integration/java/pl/allegro/tech/hermes/integration/management/QueryEndpointTest.java]: no need for `.withTrackingEnabled()`

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/query/matcher/GraterThanMatcher.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package pl.allegro.tech.hermes.management.infrastructure.query.matcher;
+
+import org.apache.commons.jxpath.JXPathException;
+import pl.allegro.tech.hermes.management.infrastructure.query.graph.ObjectGraph;
+
+public class GraterThanMatcher implements Matcher {
```
Comment (pbetkier) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/query/matcher/GraterThanMatcher.java]: That's duplicated with `LowerThanMatcher`. How about creating a `ComparisonMatcher` which would receive an operator function of (double, double) => boolean as parameter? Then implementing `GreaterThanMatcher` and `LowerThanMatcher` would just mean configuring it in a proper way, e.g.:
* greater: new ComparisonMatcher(path, parser.parseValue(node), (a, b) -> a > b);
* lower: new ComparisonMatcher(path, parser.parseValue(node), (a, b) -> a < b);

Comment (pbetkier) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/query/matcher/GraterThanMatcher.java]: typo: gr**e**ater.

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/query/matcher/MatcherInputException.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package pl.allegro.tech.hermes.management.infrastructure.query.matcher;
+
+public class MatcherInputException extends RuntimeException {
```
Comment (dankraw) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/query/matcher/MatcherInputException.java]: please extend `ManagementException`, so you'll be able to get rid of `MatcherExceptionMapper`

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/api/QueryEndpoint.java [hunk 4]
```diff
@@ -2,10 +2,7 @@
 
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
-import pl.allegro.tech.hermes.api.Group;
-import pl.allegro.tech.hermes.api.Subscription;
-import pl.allegro.tech.hermes.api.Topic;
-import pl.allegro.tech.hermes.api.Query;
+import pl.allegro.tech.hermes.api.*;
```
Comment (dankraw) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/api/QueryEndpoint.java]: IMO these wildcards are too greedy :) please tune your IDE, so explicit imports will be left instead.

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/TopicNameWithMetrics.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        return subscriptions;
+    }
+
+    public String getThroughput() {
+        return throughput;
+    }
+
+    public String getQualifiedName() {
+        return topicName.qualifiedName();
+    }
```
Comment (dankraw) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/TopicNameWithMetrics.java]: missing equals + hashcode ?

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/api/mappers/MatcherExceptionMapper.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package pl.allegro.tech.hermes.management.api.mappers;
+
+import pl.allegro.tech.hermes.api.ErrorCode;
+import pl.allegro.tech.hermes.management.infrastructure.query.matcher.MatcherInputException;
+
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.Provider;
+
+@Provider
+public class MatcherExceptionMapper extends AbstractExceptionMapper<MatcherInputException> {
```
Comment (dankraw) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/api/mappers/MatcherExceptionMapper.java]: this class can be removed, see comment for `MatcherInputException`

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicService.java [hunk 7]
```diff
@@ -215,4 +217,16 @@ public String fetchSingleMessageFromPrimary(String brokersClusterName, TopicName
                 .map(p -> new MessageTextPreview(new String(p.getContent(), StandardCharsets.UTF_8), p.isTruncated()))
                 .collect(toList());
     }
+
+    public List<TopicNameWithMetrics> getTopicsMetrics(Query<TopicNameWithMetrics> query) {
+        return query.filter(
+                getAllTopics()
+                        .stream()
+                        .map(t -> {
```
Comment (dankraw) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicService.java]: A private method doing this map() logic here would be nice so we could method-reference it

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/query/matcher/GraterThanMatcher.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    private final Object rightSideValue;
+
+    public GraterThanMatcher(String attribute, Object rightSideValue) {
+        this.attribute = attribute;
+        this.rightSideValue = rightSideValue;
+    }
+
+    @Override
+    public boolean match(Object value) {
+        try {
```
Comment (dankraw) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/query/matcher/GraterThanMatcher.java]: this try-catch is a bit wide, maybe we should divide it into smaller parts (separated graph navigation, greater-than decision logic)?

**Pull request merged**